kubernetes/updater/pkg/controller/deployment.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	"golang.org/x/mod/semver"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DeploymentVersionUpdater Reconciles a deployment by changing its img
+type DeploymentVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+func (r *DeploymentVersionUpdater) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithValues("namespacedname", req.NamespacedName)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// Get the object we are reconciling
+	var obj appsv1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
```

Comment:
Should we check if the deployment name belongs to the installation? 
Users can have multiple deployments in the same namespace and even multiple `teleport-kube-agents` installations in the same namespace under a different name. 
We must verify that the deployment is the correct one and discard if it's not the case

Reply:
I was doing it [in the manager cache configuration](https://github.com/gravitational/teleport/pull/22067/files#diff-68cb53f8985c7e8d53d1bb93fe0f27ad6d3cc65f6e00b3e1f45c804e78007232R91-R97) (by default the controller runtime caches all objects it watches, this causes OOM when running on large clusters). I can add another check in the reconciliation loop if you think it's useful.